paasta_tools/kubernetes/bin/kubernetes_remove_evicted_pods.py

```diff
@@ -0,0 +1,155 @@
+    )
+    parser.add_argument(
+        "-n", "--dry-run", action="store_true", dest="dry_run", default=False
+    )
+    args = parser.parse_args()
+    return args
+
+
+def get_evicted_pods(pods: Sequence[V1Pod]) -> Sequence[V1Pod]:
+    return [
+        pod
+        for pod in pods
+        if pod.status.phase == "Failed" and pod.status.reason == "Evicted"
+    ]
+
+
+def get_pod_service(pod: V1Pod) -> str:
+    return pod.metadata.labels.get("paasta.yelp.com/service")
+
+
+def notify_service_owners(
+    services: Mapping[str, List[EvictedPod]], soa_dir: str, dry_run: bool
+) -> None:
+    check_overrides = {
+        "page": False,
+        "alert_after": "0m",
+        "realert_every": 1,
+        "tip": "Pods can be Evicted if they go over the allowed quota for a given resource. Check the Eviction message to figure out which resource quota was breached",
+    }
+    for service in services.keys():
+        check_name = f"pod-eviction.{service}"
+        check_output = "The following pods have been evicted and will be removed from the cluster:\n"
+        for pod in services[service]:
+            check_output += f"- {pod.podname}: {pod.eviction_msg}\n"
+        if dry_run:
+            log.info(f"Would have notified owners for service {service}")
+        else:
+            log.info(f"Notifying owners for service {service}")
+            send_event(
+                service,
+                check_name,
+                check_overrides,
+                Status.CRITICAL,
+                check_output,
+                soa_dir,
+            )
+
+
+def remove_pods(
+    client: KubeClient, services: Mapping[str, List[EvictedPod]], dry_run: bool
+) -> None:
+    delete_options = V1DeleteOptions()
+    for service in services:
+        # Do not remove more than 2 pods per run
+        for pod in services[service][0:2]:
+            if dry_run:
+                log.info(f"Would have removed pod {pod.podname}")
+            else:
+                client.core.delete_namespaced_pod(
+                    pod.podname,
+                    pod.namespace,
+                    body=delete_options,
+                    grace_period_seconds=0,
+                )
+                log.info(f"Removing pod {pod.podname}")
+
+
+def evicted_pods_per_service(client: KubeClient) -> Mapping[str, List[EvictedPod]]:
+    all_pods = get_all_pods(kube_client=client, namespace="")
+    evicted_pods = get_evicted_pods(all_pods)
+    log.info(f"Pods in evicted state: {[pod.metadata.name for pod in evicted_pods]}")
+    evicted_pods_aggregated = {}
+    for pod in evicted_pods:
+        service = get_pod_service(pod)
+        if service:
+            evicted_pods_aggregated[service] = evicted_pods_aggregated.get(
+                service, []
+            ) + [
+                EvictedPod(
+                    pod.metadata.name, pod.metadata.namespace, pod.status.message
+                )
+            ]
```

Comment:
```
evicted_pods_aggregated[service].append(EvictedPod(
                    pod.metadata.name, pod.metadata.namespace, pod.status.message
                ))
```

Reply:
As discussed IRL, I'm switching to **defaultdict** to simplify this snippet.